### ISSUE TITLE
Server Display Blank, Possibly Caused by Virtual Host Issues?

### ISSUE THREAD

**Issue by scoutperry** [susanBuck/e2-fall20#24]
Sorry folks, I am still totally new to back end things. Last week I managed to get the local host to display the xampp index file through the Chrome browser but not through Firefox or Safari. Now with updating the local server I'm just getting a blank page.
<img width="1225" alt="Screen Shot 2020-09-17 at 2 49 04 PM" src="https://user-images.githubusercontent.com/52388305/93527547-090aa000-f907-11ea-8ceb-74cbbc192e58.png">
Here's my local host ip setup:
<img width="491" alt="Screen Shot 2020-09-17 at 4 27 25 PM" src="https://user-images.githubusercontent.com/52388305/93527759-57b83a00-f907-11ea-99d9-e287bec74a28.png">
And my setup for virtual hosts:
<img width="509" alt="Screen Shot 2020-09-17 at 5 04 24 PM" src="https://user-images.githubusercontent.com/52388305/93528294-1c6a3b00-f908-11ea-9d0f-90254ba3d942.png">
I ended up just pushing through to set up the production server and got this as my end product:
<img width="1228" alt="Screen Shot 2020-09-17 at 2 48 43 PM" src="https://user-images.githubusercontent.com/52388305/93528449-56d3d800-f908-11ea-8a7d-139529282cbb.png">
So now I'm trying to figure out if this last display is due to the blank page issue I had earlier or if something also went wrong in connecting it to GitHub. Thanks!

**Comment by maqboolk** [comment]
Do you have "index.php" in your practice folder?

**Comment by scoutperry** [comment]
I do

**Comment by tatthung** [comment]
Hi @maqboolk, I had this issue last week with Firefox, but fixed after updating Firefox. Not sure if our issue is the same.

**Comment by susanBuck** [comment]
Hi @scoutperry - FYI I'm looking at this issue now and will follow-up shortly with more details.

**Comment by susanBuck** [comment]
I believe the issues you're having on the local server and production server are unique, so let's address them individually:

## Problem 1: Local server
I cloned your project to my machine and set up a VirtualHost for `practice.loc` to point to it. I was able to get it to load, so this confirms there's nothing wrong within the code itself - there's an index.php file, it has content, etc.

![image](https://user-images.githubusercontent.com/237711/93612307-5c2f3200-f99d-11ea-81f3-5530e0a6a01a.png)

I don't see any issues in the VirtualHost config you shared, but to be sure, can you copy/paste it exactly as you have it here in your reply? This way I can copy/paste it on my end to test - sometimes there's a weird character bug that's not visible on first glance. If I can test it on my end, we can confirm there's no bug.

Finally, let's confirm that your document root itself is in the correct location; run the following command to output the contents of your practice index.php file:

```
$ cat /Applications/XAMPP/xamppfiles/htdocs/e2/practice/index.php
```

Let me know what that yields.



## Problem 2: Production server
I see that your domain `scoutperry.me` correctly points to your DigitalOcean IP address, `157.230.62.33`. This is further confirmed by the fact that when you visit <http://scoutperry.me> we see the DigitalOcean starter page.

However, if we visit <http://practice.scoutperry.me> we see a message telling us it can't find an IP address for that address:
![image](https://user-images.githubusercontent.com/237711/93611553-73b9eb00-f99c-11ea-9d17-801e9b6deb59.png)

Can you share a screenshot of your domain's DNS settings so we can confirm they were set up correctly? Either there's an error there, *or* there's also a possibility that the DNS settings are just stuck in a cache and haven't propagated yet. Either way, double checking the settings is a good place to start.

**Comment by scoutperry** [comment]
Ok so here's how I have the Virtual host info:
```
<VirtualHost *:80>
    ServerName practice.loc
    DocumentRoot /Applications/XAMPP/xamppfiles/htdocs/e2/practice
    <Directory /Applications/XAMPP/xamppfiles/htdocs/e2/practice>
        AllowOverride All
        Options -Indexes
        Require all granted
    </Directory>
</VirtualHost>

<VirtualHost *:80>
    ServerName localhost
    DocumentRoot /Applications/XAMPP/xamppfiles/htdocs
    <Directory /Applications/XAMPP/xamppfiles/htdocs>
        AllowOverride All
        Options -Indexes
        Require all granted
    </Directory>
</VirtualHost>

<VirtualHost *:80>
    ServerName e2p1.loc
    DocumentRoot /Applications/XAMPP/xamppfiles/htdocs/e2/p1
    <Directory /Applications/XAMPP/xamppfiles/htdocs/e2/p1>
        AllowOverride All
        Options -Indexes
        Require all granted
    </Directory>
</VirtualHost>
```
And here's the results of the cat command:
```
$ cat /Applications/XAMPP/xamppfiles/htdocs/e2/practice/index.php
<!doctype html>
<html lang='en'>
<head><title>Practice Application</title>
    <meta charset='uft-8'>
</head>
<body>
    <h1>Practice Application</h1>
<p>hello<p>
</body>
</html>
```
Sorting out problem 2 next

**Comment by scoutperry** [comment]
Here are the DNS settings:
<img width="1225" alt="Screen Shot 2020-09-18 at 1 59 57 PM" src="https://user-images.githubusercontent.com/52388305/93630224-a32a2100-f9b7-11ea-86e8-9fb54b070b46.png">

**Comment by scoutperry** [comment]
lol is it the lack of asterisk????

**Comment by scoutperry** [comment]
<img width="1228" alt="Screen Shot 2020-09-18 at 2 06 20 PM" src="https://user-images.githubusercontent.com/52388305/93630587-3a8f7400-f9b8-11ea-8eb8-8eeab7d665a6.png">
OK, Changed that.

**Comment by susanBuck** [comment]
Hi @scoutperry -

As of Saturday AM, it looks like <http://practice.scoutperry.me> is working now, so we can check the production issue off the list.  :)

Back to the local server problem..

I tested the VirtualHosts configs you shared on my local server, and they worked without issue - so we can assume the problem does not lie there.

Another thing to try...

Can you run the command `/Applications/XAMPP/bin/httpd -S` and share the full output?

This will give us some debugging information about what Apache is seeing ([ref](https://hesweb.dev/e2/notes/local-server/local-domains#apache-wont-start-after-editing-virtualhosts-file-codehttpd-vhostsconfcode))

Specifically, we want to confirm it is reading the configs from your `/Applications/XAMPP/xamppfiles/etc/extra/httpd-vhosts.conf` file.

For reference, here's what the output of that command looks like on my system:

```
/Applications/XAMPP/xamppfiles/htdocs   $ /Applications/XAMPP/bin/httpd -S
AH00112: Warning: DocumentRoot [/Applications/XAMPP/xamppfiles/docs/dummy-host.example.com] does not exist
AH00112: Warning: DocumentRoot [/Applications/XAMPP/xamppfiles/docs/dummy-host2.example.com] does not exist
VirtualHost configuration:
*:443                  is a NameVirtualHost
         default server hesweb.loc (/Applications/XAMPP/xamppfiles/etc/extra/httpd-vhosts.conf:155)
         port 443 namevhost hesweb.loc (/Applications/XAMPP/xamppfiles/etc/extra/httpd-vhosts.conf:155)
         port 443 namevhost courseintros.loc (/Applications/XAMPP/xamppfiles/etc/extra/httpd-vhosts.conf:170)
*:80                   is a NameVirtualHost
         default server dummy-host.example.com (/Applications/XAMPP/xamppfiles/etc/extra/httpd-vhosts.conf:23)
         port 80 namevhost dummy-host.example.com (/Applications/XAMPP/xamppfiles/etc/extra/httpd-vhosts.conf:23)
                 alias www.dummy-host.example.com
         port 80 namevhost dummy-host2.example.com (/Applications/XAMPP/xamppfiles/etc/extra/httpd-vhosts.conf:32)
         port 80 namevhost practice.loc (/Applications/XAMPP/xamppfiles/etc/extra/httpd-vhosts.conf:40)
         port 80 namevhost e28p1.loc (/Applications/XAMPP/xamppfiles/etc/extra/httpd-vhosts.conf:61)
         port 80 namevhost e2p1.loc (/Applications/XAMPP/xamppfiles/etc/extra/httpd-vhosts.conf:72)
         port 80 namevhost student.loc (/Applications/XAMPP/xamppfiles/etc/extra/httpd-vhosts.conf:86)
         port 80 namevhost susanbuck.loc (/Applications/XAMPP/xamppfiles/etc/extra/httpd-vhosts.conf:198)
         port 80 namevhost localhost (/Applications/XAMPP/xamppfiles/etc/extra/httpd-vhosts.conf:227)
ServerRoot: "/Applications/XAMPP/xamppfiles"
Main DocumentRoot: "/Applications/XAMPP/xamppfiles/htdocs"
Main ErrorLog: "/Applications/XAMPP/xamppfiles/logs/error_log"
Mutex default: dir="/Applications/XAMPP/xamppfiles/logs/" mechanism=default 
Mutex mpm-accept: using_defaults
Mutex authdigest-opaque: using_defaults
Mutex proxy-balancer-shm: using_defaults
Mutex ssl-stapling-refresh: using_defaults
Mutex rewrite-map: using_defaults
Mutex authdigest-client: using_defaults
Mutex ssl-stapling: using_defaults
Mutex ldap-cache: using_defaults
Mutex proxy: using_defaults
Mutex ssl-cache: using_defaults
Mutex authn-socache: using_defaults
PidFile: "/Applications/XAMPP/xamppfiles/logs/httpd.pid"
Define: DUMP_VHOSTS
Define: DUMP_RUN_CFG
Define: MODPERL2
User: name="Susan" id=502 not_used
Group: name="daemon" id=1 not_used
```

**Comment by susanBuck** [comment]
Follow-up as I was thinking about this more:

Do you also see a blank white page when you visit <http://localhost>?

**Comment by maqboolk** [comment]
> Follow-up as I was thinking about this more:
> 
> Do you also see a blank white page when you visit http://localhost?

@susanBuck  Hi Susan , I am not seeing blank page. Thank you

**Comment by susanBuck** [comment]
@maqboolk - My reply was to @scoutperry. Are you having issues on your local server as well? If so, feel free to create a new thread and I can help you troubleshoot there. : )

**Comment by maqboolk** [comment]
@susanBuck 
I have no issues with the local server.

**Comment by scoutperry** [comment]
@susanBuck Sorry for the delay! Using `$ /Applications/XAMPP/bin/httpd -S` I got the following:

```
AH00112: Warning: DocumentRoot [/Applications/XAMPP/xamppfiles/docs/dummy-host.example.com] does not exist
AH00112: Warning: DocumentRoot [/Applications/XAMPP/xamppfiles/docs/dummy-host2.example.com] does not exist
VirtualHost configuration:
*:80                   is a NameVirtualHost
         default server dummy-host.example.com (/Applications/XAMPP/xamppfiles/etc/extra/httpd-vhosts.conf:23)
         port 80 namevhost dummy-host.example.com (/Applications/XAMPP/xamppfiles/etc/extra/httpd-vhosts.conf:23)
                 alias www.dummy-host.example.com
         port 80 namevhost dummy-host2.example.com (/Applications/XAMPP/xamppfiles/etc/extra/httpd-vhosts.conf:32)
         port 80 namevhost practice.loc (/Applications/XAMPP/xamppfiles/etc/extra/httpd-vhosts.conf:40)
         port 80 namevhost localhost (/Applications/XAMPP/xamppfiles/etc/extra/httpd-vhosts.conf:50)
         port 80 namevhost e2p1.loc (/Applications/XAMPP/xamppfiles/etc/extra/httpd-vhosts.conf:60)
ServerRoot: "/Applications/XAMPP/xamppfiles"
Main DocumentRoot: "/Applications/XAMPP/xamppfiles/htdocs"
Main ErrorLog: "/Applications/XAMPP/xamppfiles/logs/error_log"
Mutex ssl-stapling-refresh: using_defaults
Mutex rewrite-map: using_defaults
Mutex authdigest-client: using_defaults
Mutex ssl-stapling: using_defaults
Mutex ldap-cache: using_defaults
Mutex proxy: using_defaults
Mutex ssl-cache: using_defaults
Mutex authn-socache: using_defaults
Mutex default: dir="/Applications/XAMPP/xamppfiles/logs/" mechanism=default 
Mutex mpm-accept: using_defaults
Mutex authdigest-opaque: using_defaults
Mutex proxy-balancer-shm: using_defaults
PidFile: "/Applications/XAMPP/xamppfiles/logs/httpd.pid"
Define: DUMP_VHOSTS
Define: DUMP_RUN_CFG
Define: MODPERL2
User: name="daemon" id=1 not_used
Group: name="daemon" id=1 not_used
```
Both localhost and practice.loc are blank. To add another wrinkle, if I use any browser other than chrome, it will just straight up say 'Server not found.' Thank you for your help, even a partial fix is a win

**Comment by susanBuck** [comment]
The output you shared suggests your VirtualHosts file is "healthy" and practice.loc *should* be mapping correctly.

You can try running the command `ping practice.loc` in command line to confirm your hosts file is properly routing traffic to this domain to your local server at 127.0.0.1.

The fact that you see a blank white page in Chrome leads me to think it's some code/permissions error.
But the fact that you see _"Server not found"_ on other browsers leads me to think it's a server config problem.

In short, this is a head scratcher. : )

Do you have time tomorrow (Tuesday) to do a Zoom screenshare so we can debug this real-time? Let me know when would work for you - I'm available from Noon eastern on.

**Comment by scoutperry** [comment]
Would sometime this evening work? Or I can do after 3pm tomorrow

**Comment by susanBuck** [comment]
@scoutperry - I could do 8:30pm Eastern this evening. Or 3pm tomorrow also works. Whatever works best for you.

**Comment by scoutperry** [comment]
8:30pm works great, thank you!

**Comment by susanBuck** [comment]
Sounds good - my zoom url is https://harvard.zoom.us/my/susanbuck

**Comment by susanBuck** [comment]
Hi @scoutperry - FYI I just realized I'm having an issue with my zoom URL that I shared above. Try this instead: 

https://zoom.us/j/9223500980?pwd=aVN0dTl5ZE9NZjI4UEovZXBPMGdSQT09

Meeting ID: 922 350 0980
Passcode: vTYm7N

**Comment by susanBuck** [comment]
Some follow-up notes on this from my Zoom meeting with Scout -

We narrowed this down to a port conflict issue. It appeared that Mac OS had an existing install of Apache that was running as a process on Scout's computer, using port 80, causing XAMPP's Apache to fail to run. 

Because of this, when Scout visited the local server in their browser, they were actually accessing the document root for the Mac-Apache, which is why there was a blank page.

When we attempted to stop the Mac-Apache process via the *Activity Monitor* ([as described in the notes...](https://hesweb.loc/e2/notes/local-server/xampp-mac) it was unsuccessful because the process kept re-generating itself.

After doing some Googling, I found the following command to kill and disable this process:

```
$ sudo launchctl unload -w /System/Library/LaunchDaemons/org.apache.httpd.plist
```

Also, to be sure, we ran this command to kill any other rogue Apache processes:

```
$ sudo killall httpd
```

After we did that, XAMPP's Apache successfully started. 

Some additional notes/points of reference:

The following command can be used to get a list of any/all processes using port 80 on your mac:

```
$ sudo lsof -i :80
```

If this finds any processes, it will list an id for that process in the left column. 

You can then use the following command to find out information about that process (using process id of 12345 as an example):

```bash
$ ps u 12345
```

And you can use this command to kill a process:

```bash
$ kill 12345
```
